src/env-inl.h

```diff
@@ -101,6 +101,20 @@ inline void Environment::AsyncHooks::set_enable_callbacks(uint32_t flag) {
   fields_[kEnableCallbacks] = flag;
 }
 
+inline Environment::AsyncCallbackScope::AsyncCallbackScope(Environment* env)
+    : env_(env) {
+  env_->async_callback_cntr_++;
+}
+
+inline Environment::AsyncCallbackScope::~AsyncCallbackScope() {
+  env_->async_callback_cntr_--;
+  CHECK_GE(env_->async_callback_cntr_, 0);
+}
+
+inline bool Environment::AsyncCallbackScope::in_tick() {
```

Comment:
> in_tick is meant to indicate that we're currently "in" the state of processing the next "tick" queue. 

Right, but the member variable `async_callback_cntr`, which I interpret as "the nesting level of this AsyncCalbackScope instance" doesn't _in itself_ indicate whether the next "tick" queue is being processed, does it? It seems that it's more that we want to process that next "tick" queue only once per set of reentrant calls to `MakeCallback`, at the end of the first/root `MakeCallback` call, and we know that we're completing that call if `async_callback_cntr` is `== 1`. Or am I missing something?

> I don't have a code sample that can reproduce that, possibly @indutny might, but that call itself isn't recursive.

It's reentrant but not recursive, so that mean the call to MakeCallback in that case can be interrupted by another call to MakeCallback?


> It's that from JS we ask the HTTPParser for more header information. Which would mean MakeCallback would be called twice. 

OK, do you have a link to where that is done in code?

> While at the same time it's possible for the HTTPParser to initiate the JS stack.

What do you mean by "initiate the JS stack"?

My apologies for the annoying questions, I'm mostly trying to make sure I understand all the implications of this PR :)